.clang-format

```diff
@@ -0,0 +1,138 @@
+---
+Language:        Cpp
+# BasedOnStyle:  Mozilla
+AccessModifierOffset: -3
+AlignAfterOpenBracket: Align
+AlignArrayOfStructures: Left
+AlignConsecutiveMacros: false
+AlignConsecutiveAssignments: true
+AlignConsecutiveDeclarations: true
+AlignEscapedNewlines: Right
+AlignOperands:   true
+AlignTrailingComments: true
+AllowAllArgumentsOnNextLine: true
+AllowAllConstructorInitializersOnNextLine: true
+AllowAllParametersOfDeclarationOnNextLine: false
+AllowShortBlocksOnASingleLine: Empty
+AllowShortCaseLabelsOnASingleLine: true
+AllowShortFunctionsOnASingleLine: Inline
+AllowShortLambdasOnASingleLine: All
+AllowShortIfStatementsOnASingleLine: Never
+AllowShortLoopsOnASingleLine: false
+AlwaysBreakAfterDefinitionReturnType: None
+AlwaysBreakAfterReturnType: None
+AlwaysBreakBeforeMultilineStrings: false
+AlwaysBreakTemplateDeclarations: Yes
+BinPackArguments: false
+BinPackParameters: false
+BraceWrapping:
+  AfterCaseLabel:  false
+  AfterClass:      false
+  AfterControlStatement: Never
+  AfterEnum:       false
+  AfterFunction:   false
+  AfterNamespace:  false
+  AfterObjCDeclaration: false
+  AfterStruct:     false
+  AfterUnion:      false
+  AfterExternBlock: false
+  BeforeCatch:     false
+  BeforeElse:      false
+  IndentBraces:    false
+  SplitEmptyFunction: false
+  SplitEmptyRecord: false
+  SplitEmptyNamespace: true
+BreakBeforeBinaryOperators: None
+BreakBeforeBraces: Custom
+BreakBeforeInheritanceComma: false
+BreakInheritanceList: BeforeComma
+BreakBeforeTernaryOperators: true
+BreakConstructorInitializersBeforeComma: false
+BreakConstructorInitializers: BeforeComma
+BreakAfterJavaFieldAnnotations: false
+BreakStringLiterals: true
+ColumnLimit:     120
+CommentPragmas:  '^ IWYU pragma:'
```

Comment:
I don't think it's just about the modern displays supporting the larger resolutions. The longer the column width the smaller the characters become, so it gets increasingly harder for the human eye to parse longer lines of code. As a developer you also want to avoid or minimize sideways scrolling. Also wrapping text is fine if you're reading prose, but wrapping lines of code is quite undesirable for a developer who's already trying to keep up with scope as he reads/implements code.

Reply:
> I don't think it's just about the modern displays supporting the larger resolutions. The longer the column width the smaller the characters become, so it gets increasingly harder for the human eye to parse longer lines of code. As a developer you also want to avoid or minimize sideways scrolling. 

That is very true. Also many developers like to have multiple editors side-by-side, and a reasonable column width helps.

Reply:
Characters are a fixed size, they do not get "smaller" as line length increases. The readability of line wrapping and "desirability" of sideways scrolling is subjective. Are there any objective reasons to have a line length limit? Is this worth failing a build or rejecting a PR over?

Reply:
> Characters are a fixed size, they do not get "smaller" as line length increases.

Obviously. Did I suggest otherwise?

> Are there any objective reasons to have a line length limit?

There are reasons to have a line width limit I stated below. Whether they are objective or subjective may reside in the eyes of the beholder. In any case, that are **not** any more subjective than your arguments against it.

It seems to me that you are subject to the status quo bias, in the sense that you feel that either:
- you absolutely prove that a fixed line width is better
- or we should remain in the current state and not enforce a line width.

Instead, we should consider the benefits and drawback of limiting the line width or not (remember this PR doesn't require enforcement), and decide whether setting a limit would overall benefit our code quality and developer productivity.

I'm getting very strong reactions on this PR. It may be because some have not used clang-format before, and are weary on additional restriction on their coding style. I think this is very understandable, but I know from experience that, instead of a straightjacket, using clang-format is actually freeing for the developer, as it takes care of the formatting details and allows us to focus on what matters, the semantics of the code we write.

Reply:
Line length limits:
**Pros**
1. A minority of contributors, using specific editors with specific window configurations, do not have to scroll sideways or view wrapped lines, which is their personal preference but does not otherwise impact code correctness or quality.
1. Better Version Control: Limiting the line length can reduce the number of changes in a single line, which can make it easier to compare different versions of the code using version control tools.

**Cons**
1. Inflexibility: Setting a strict line length limit can sometimes lead to code that is difficult to read, especially when working with long function or variable names.
1. Reduced Expressiveness: Limiting the line length can also reduce the expressiveness of the code, making it harder to express complex concepts and algorithms in a concise manner.
1. Increased Verbosity: In order to conform to the line length limit, developers may need to add additional line breaks or whitespace, leading to increased verbosity in the code.
1. Tyranny of the minority.
1. The "undesired" line wrapping of an editor is traded for lines wrapped _for everyone_, regardless of their number of available columns.
1. Code is truncated, making the small minority of use-cases where lines longer than 120 char _are_ appropriate impossible to implement, including but not limited to logic statements with nested parenthesis.
1. Comments are truncated, making useful inline comments at the end of lines with a significant amount of code impossible...such as the kinds of lines where these comments would be most appropriate.
1. The minority of contributors who prefer inline comments _not_ to be wrapped have to put up with it to satisfy this other minority of contributors who prefer arbitrary line length limits.
1. Tooling relying on parsing comments is broken, such as [doxygen](https://www.startpage.com/do/search?q=doxygen+and+line+length+limits&segment=startpage.brave).
1. Comments referencing external URIs with further explanation about implementation details are now broken, as URIs within comments are also subject to length limits.
1. When developers or engineers encounter edge-cases that require lines longer than the limit, they are [forced to disable **_all_** `clang-format` rules](https://stackoverflow.com/a/53005855) in order to bypass the line length limit, preventing `clang-format` from protecting developers from other errors.

After making a pro/con list, it is clear the line length limit should be set to zero.

Reply:
> Characters are a fixed size, they do not get "smaller" as line length increases. The readability of line wrapping and "desirability" of sideways scrolling is subjective.

I was referring to the character relative size on the screen w.r.t. the column width. If you increase the column width to fit more visible characters, those characters will appear smaller on the screen as compared to a smaller column width. In my opinion, this is a very strong, objective reason to limit the column width. If we allow it to be arbitrarily long, code readability, scalability, and maintenance will be affected. This is independent of the character absolute size in pixels, which of course, won't change.

Reply:
Yes, if there are more columns then each column represents a smaller portion of the total characters displayed horizontally, I guess...I fail to see how this is a disadvantage, or how imposing line length limits is preferable to the cons listed above.

Reply:
I just read your list of cons and realize that most of the disadvantages you point out stem from the 120 characters limit, not from the actual enforcement of the rule. So maybe 120 is too restrictive and should be increased to 130, 140, 150?

I could also argue against some of your cons by pointing out that most variable names used in our code base don't really use long compound names, that longer code lines might actually obfuscate logic and semantics further by making unnecessarily long constructs instead of encouraging more incremental code with more descriptive variables that dissect the logic better, that code expressiveness is rather achieved and improved through better code factorization practices not through longer line constructs, etc.

Lastly, I'm really curious to know whether the current 120 character limit that we use in our C++ code base, albeit not enforced automatically, has really been that restrictive to our developers and contributors re: the cons that you list. Has it really made the code more inflexible, harder to read, reduced expressiveness, or created more verbosity as a result? If the answer is yes to some of the above, then maybe this discussion should probably be about increasing the column width limit while still allowing the rule to be enforced on the code. Btw, I think the rule should apply to code only, not to multiline comments, although I'm not sure .clang-format can make that distinction.

Reply:
> I just read your list of cons and realize that most of the disadvantages you point out stem from the 120 characters limit, not from the actual enforcement of the rule. So maybe 120 is too restrictive and should be increased to 130, 140, 150?

Only con number six referenced a specific number, and it would be applicable to _any_ of the other numbers you suggested. I have edited it to replace "120" with `n` to be more clear. All of these cons still stand as I see them.

> I **_could_** also argue against **_some_** of your cons by pointing out that **_most_** variable names used in our code base don't really use long compound names, that longer code lines **_might_** actually obfuscate logic and semantics further[...]

Emphasis mine.

This is very slippery language. If you have specific, productive dissent to my pro/con list, please share it without speaking in the hypothetical.

I observe you used terms like "some," "most," and "might" to carefully side-step the very edge-cases I am deeply concerned about. The fact that you felt the need to use these modifiers proves my point - line length limits **_need_** to be left to the discretion of the developers and the peer reviewers, not enforced programmatically. And I say that as the Automation Engineer...it is in my personal best-interest for my customers to wish to enforce these kinds of things programmatically because that is job security for me. However, it is not in the best interest of the Leap project, or our other projects.

Take, for example, con number eleven. If you believe that **_any_** of the other rules are important, then we do not wish to create circumstances where developers are incentived to disable `clang-format`.

> I'm really curious to know whether the current 120 character limit that we use in our C++ code base[...]

We do not currently use a line length limit in our codebase. We never have, not in the five years that EOSIO/Antelope has existed. If we already have a length limit implemented, _why are you advocating for a second one_?

I have asked this question several times already, and I will ask it again.  
**What are specific examples of cases where the absence of a line length limit cost the ENF time or money?**  
Nobody has even acknowledged this question, let alone provided specific examples. This suggests, to me, that we are trying to solve a _hypothetical_ problem instead of a _real_ problem.

@greg7mdp set a bar...

> we should consider the benefits and drawback of limiting the line width or not [...], and decide whether setting a limit would overall benefit our code quality and developer productivity.

...and I cleared it. Rather than trying to invert the discussion, please show your ideas are justified, can be defended, and withstand scrutiny.